### DISCUSSION TITLE
To use rdkit.Chem.Fragments module to extract a substructure indices

### DISCUSSION THREAD

**Discussion by zmahnoor14** [rdkit/rdkit#7503] [General]
Hi RDKit community,

thanks for the awesome tool : )

So, I am using rdkit.Chem.Fragments module to detect functional groups in my molecules. Let's take an example of amide functional group.

```python
smile_secondaryAmide = "CN(CCCNC(=O)C1CCCO1)C2=NC3=CC(=C(C=C3C(=N2)N)OC)OC"
mol1 = Chem.MolFromSmiles(smile_secondaryAmide)
Chem.Fragments.fr_amide(mol1)
```
This code returns that my molecule has 1 amide functional group. Now, if I want to extract the substructure of this amide, I simply use the following code:

```python
amide_smiles = "*C(N*)=O"
amide_substructure = Chem.MolFromSmarts(amide_smiles)
mol1.GetSubstructMatches(amide_substructure)
```

This code gives me the indices: ((8, 6, 5, 4, 7),) PERFECT!!!!

The problem arises when I have a primary amide in my list of amides. Here I use the same code on the primary amide structure:

```python
smile_primaryAmide = "CC(C)NCC(COC1=CC=C(C=C1)CC(=O)N)O"
mol2 = Chem.MolFromSmiles(smile_primaryAmide)
Chem.Fragments.fr_amide(mol2)
```
This code returns that my molecule has 1 amide functional group. BUT, the mol1.GetSubstructMatches function doesnt detect this with the smiles `"*C(N*)=O"`, rather `"*C(N)=O"`.

```python
# with "*C(N*)=O" gives no indices
amide_smiles = "*C(N*)=O"
amide_substructure = Chem.MolFromSmarts(amide_smiles)
mol2.GetSubstructMatches(amide_substructure)
# output is ()

# with "*C(N)=O" gives no indices
amide_smiles = "*C(N)=O"
amide_substructure = Chem.MolFromSmarts(amide_smiles)
mol2.GetSubstructMatches(amide_substructure)
# output is ((14, 15, 17, 16),)
```

There could be some solutions for this; but I was wondering, if rdkit.Chem.Fragments can be directly used to extract the indices as well along with the number of functional groups. Could anyone let me know if this is possible? Or an easier solution to this problem? I also couldn't find the source code for rdkit.Chem.Fragments in the github? If someone can direct me to the link, that would also be great

Goal: I would like to write one function that detects any kind of amide and gives me the indices; instead of writing a separate condition for the primary amide.

Any suggestions would be very appreciated. Thank you!

Kind regards,
Mahnoor

**Comment by ruslankotl** [comment]
Dear Mahnoor,

This might be a late answer, but the problem lies in how you define your SMARTS for amides
`*` denotes a dummy atom, i.e., any heavy atom. In case of `C(N*)=O`, SMARTS will not match primary amides because it expects two heavy atoms connected to nitrogen: the carbonyl carbon and any other heavy atom.
This expression `[#6]-[C;X3]([N;X3;+0])=[O;X1]` should match all carboxylic acid amides, and also N-acyl hydroxylamines etc.
This expression `[#6]-[C;X3]([N;X3;+0;$([*;H2]C),$([*;H1](C)[#6]),$([*;H0](C)([#6])[#6])])=[O;X1]` is even more complicated, but should ideally only cover amides and nothing else.


I personally recommend using SMARTS official specification [https://www.daylight.com/dayhtml/doc/theory/theory.smarts.html](https://www.daylight.com/dayhtml/doc/theory/theory.smarts.html) and Smarter SMARTS online tool [https://baoilleach.github.io/smartersmarts-js/#search/openbabel/[%236]-[C%3BX3]([N%3BX3%3B%2B0])%3D[O%3BX1]/](https://baoilleach.github.io/smartersmarts-js/#search/openbabel/[%236]-[C%3BX3]([N%3BX3%3B%2B0])%3D[O%3BX1]/) for desigining and refining your queries.
